packages/core/primitives/event-dispatch/src/event_dispatcher.ts

```diff
@@ -0,0 +1,153 @@
+            this.dispatchToDelegate(eventInfoWrapper);
+          }
+        },
+      },
+    );
+  }
+
+  /**
+   * The entrypoint for the `EventContract` dispatch.
+   */
+  dispatch(eventInfo: EventInfo): void {
+    this.dispatcher.dispatch(eventInfo);
+  }
+
+  /** Internal method that does basic disaptching. */
+  private dispatchToDelegate(eventInfoWrapper: EventInfoWrapper) {
+    prepareEventForBubbling(eventInfoWrapper);
+    while (eventInfoWrapper.getAction()) {
+      prepareEventForDispatch(eventInfoWrapper);
+      this.dispatchDelegate(eventInfoWrapper.getEvent(), eventInfoWrapper.getAction()!.name);
+      if (propagationStopped(eventInfoWrapper)) {
+        return;
+      }
+      this.actionResolver.resolveParentAction(eventInfoWrapper.eventInfo);
+    }
+  }
+}
+
+function prepareEventForBubbling(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  const stopPropagation = () => {
+    event[PROPAGATION_STOPPED_SYMBOL] = true;
+  };
+  Object.defineProperty(event, 'stopPropagation', {
+    value: stopPropagation,
+  });
+  Object.defineProperty(event, 'stopImmediatePropagation', {
+    value: stopPropagation,
+  });
+}
+
+function propagationStopped(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  return !!event[PROPAGATION_STOPPED_SYMBOL];
+}
+
+function prepareEventForReplay(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  const target = eventInfoWrapper.getTargetElement();
+  Object.defineProperty(event, 'target', {
+    value: target,
+  });
+  Object.defineProperty(event, 'eventPhase', {
+    value: EventPhase.REPLAY,
+  });
+  if (ngDevMode) {
+    Object.defineProperty(event, 'preventDefault', {
+      value: () => {
+        throw new Error(
+          '`preventDefault` called during event replay. Because event replay occurs after browser ' +
+            'dispatch, `preventDefault` would have no effect. You can check whether an event is ' +
+            'being replayed by accessing the event phase: `event.eventPhase === EventPhase.REPLAY`.',
+        );
+      },
+    });
+    Object.defineProperty(event, 'composedPath', {
+      value: () => {
+        throw new Error(
+          '`composedPath` called during event replay. Because event replay occurs after browser ' +
+            'dispatch, `composedPath()` will be empty. Iterate parent nodes from `event.target` or ' +
+            '`event.currentTarget` if you need to check elements in the event path.',
+        );
+      },
+    });
+  }
+}
+
+function prepareEventForDispatch(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  const currentTarget = eventInfoWrapper.getAction()?.element;
+  if (currentTarget) {
+    Object.defineProperty(event, 'currentTarget', {
+      value: currentTarget,
+      configurable: true, // `currentTarget` is going to get reassigned every dispatch.
+    });
```

Comment:
As an idea: we can create a separate "patching" function, which would also contain information and context on why we do this (since an event instance gets changed after being dispatched in a browser).
```suggestion
/** (explain why we do patching) */
function patchEventInstance<T>(event: Event, property: string, value: T, configurable = false) {
  Object.defineProperty(event, property, { value, configurable });
}

    // ... later in the code ...
    const configurable = true; // `currentTarget` is going to get reassigned every dispatch.
    patchEventInstance(event, 'currentTarget', currentTarget, configurable);
```

Reply:
Done. Let me know if you have any edits for my explanation.